Uniform random numbers (refactored)

TimHanel00 opened the pull request:
Adds Philox random uniform Engines (under alpaka::rand -- support for both host and device execution)
Adds uniformReal distribution mirroring the std::uniform_real interface (constexpr type dispatch and Interval selection).
Adds a dedicated example (randomInit) and a unit test for uniformReal distributions and philox engines.

mehmetyusufoglu committed:
copy random functions from alpaka

mehmetyusufoglu committed:
only take philox and create related example

TimHanel00 committed:
UniformReal refactor (supports double + float and Intervals), folder and namespace restructuring, and add a uniformReal Test

psychocoderHPC commented on example/randomInit/src/randomInit.cpp:
```diff
@@ -0,0 +1,404 @@
+            {
+                // Generate numbers in [0,1). Philox engine already generates uniform random numbers but they are
+                // integers.
+                alpaka::rand::distribution::UniformReal<float> dist;
```
please check if the interface is fitting the interval mentioned in the comments above `[0, 1)`

TimHanel00 replied:
checked that: CO [0,1) is the default interval.

psychocoderHPC commented on include/alpaka/rand/distribution/interval.hpp:
```diff
@@ -0,0 +1,31 @@
+    /// **Interval type (a, b]: open (exclusive) at the lower bound, closed (inclusive) at the upper bound.**
+    struct OC
+    {
+        constexpr OC() = default;
```
Please remove, not nessesary.

psychocoderHPC commented on include/alpaka/rand/distribution/interval.hpp:
```diff
@@ -0,0 +1,31 @@
+    {
+        constexpr CO() = default;
+    };
+
```
please add `constexpr auto co = CO{}`

psychocoderHPC commented on include/alpaka/rand/distribution/uniformReal.hpp:
```diff
@@ -0,0 +1,349 @@
+    /** Returns a constant, which is equivalent to std::nextafter(T_Floating(1), T_Floating(0))
+     * or more specifically: returns the highest floating-point value lower than one */
+    template<std::floating_point T>
+    consteval T prev_one() noexcept
```
should be `prevOne`, we use camelCase.

psychocoderHPC commented on include/alpaka/rand/distribution/uniformReal.hpp:
```diff
@@ -0,0 +1,349 @@
+     * matches. It should never be instantiated and exists only to
+     * catch unsupported interval configurations.
+     */
+    template<typename T_Engine, typename T_Interval, std::integral T_Integer, std::floating_point T_Floating>
```
you said T_Interval need concept instead of typename

psychocoderHPC commented on include/alpaka/rand/distribution/uniformReal.hpp:
```diff
@@ -0,0 +1,349 @@
+    template<typename T_Engine, std::integral T_Integer, typename T_Floating>
+    struct IntervalAwareConversion<T_Engine, interval::CO, T_Integer, T_Floating>
+    {
+        static constexpr T_Floating normalizedInterval
```
if  it should not be exposed please move it without `static` into the method

psychocoderHPC commented on include/alpaka/rand/distribution/uniformReal.hpp:
```diff
@@ -0,0 +1,349 @@
+        static constexpr T_Floating normalizedInterval
+            = prev_one<T_Floating>() / static_cast<T_Floating>(std::numeric_limits<T_Integer>::max());
+
+        ALPAKA_FN_HOST_ACC auto operator()(T_Integer const& i) const
```
remove`ALPAKA_FN_HOST_ACC` by `constexpr`, only if `nvcc` complains than use `ALPAKA_FN_HOST_ACC constexpr`

psychocoderHPC commented on include/alpaka/rand/distribution/uniformReal.hpp:
```diff
@@ -0,0 +1,349 @@
+
+        ALPAKA_FN_HOST_ACC auto operator()(T_Integer const& i) const
+        {
+            return static_cast<T_Floating>(i) * normalizedInterval + normalizedInterval / T_Floating(2.0);
```
should be `T_Floating{2.0}`

psychocoderHPC commented on include/alpaka/rand/distribution/uniformReal.hpp:
```diff
@@ -0,0 +1,349 @@
+    };
+
+    /** Convert an integer RNG output to a floating-point type in the interval (0, 1].
+     *   * **Inspired by NVIDIA cuRAND's approach to avoid generating zero.
```
something is wrong with `**`

psychocoderHPC commented on include/alpaka/rand/distribution/uniformReal.hpp:
```diff
@@ -0,0 +1,349 @@
+     * This is the default case where the engine result type already matches and thus the engine is simply invoked.
+     */
+    template<typename T_Engine, uint32_t byteLengthEngineResult, uint32_t byteLengthRealType>
+    struct bitLengthConformityAdapter
```
should be `BitLengthConformityAdapter`

psychocoderHPC commented on include/alpaka/rand/distribution/uniformReal.hpp:
```diff
@@ -0,0 +1,349 @@
+            "Result returned by the randomBitGenerator does not have a length that is accepted by the uniformReal "
+            "distribution!");
+        static_assert(
+            (byteLengthEngineResult == 4u || byteLengthRealType == 8u),
```
bug should be `8 and 4`

psychocoderHPC commented on include/alpaka/rand/distribution/uniformReal.hpp:
```diff
@@ -0,0 +1,349 @@
+            byteLengthEngineResult == byteLengthRealType,
+            "By logic this should never fail in case the compiler accepts the specialization of the adapter!");
+
+        ALPAKA_FN_HOST_ACC auto operator()(T_Engine& engine)
```
the method can maybe be `const`, please check

psychocoderHPC commented on include/alpaka/rand/engine/philox/multiplyAndSplit64to32.hpp:
```diff
@@ -0,0 +1,43 @@
+     * @param resultLow low 32 bits of the product a*b
+     */
+    // TODO: See single-instruction implementations in original Philox source code
+    inline constexpr void multiplyAndSplit64to32(
```
is inline here required, my IDE is saying if `constexpr` is used inline is not required.

psychocoderHPC commented on include/alpaka/rand/engine/philox/PhiloxBaseCommon.hpp:
```diff
@@ -0,0 +1,113 @@
+        static ALPAKA_FN_HOST_ACC void advanceCounter(alpaka::Vec<T, N>& counter)
+        {
+            ++counter[0];
+
```
could you remove the empty lines

psychocoderHPC commented on include/alpaka/rand/engine/philox/PhiloxSingle.hpp:
```diff
@@ -0,0 +1,126 @@
+        ALPAKA_FN_HOST_ACC auto nextNumber()
+        {
+            // Element zero will always contain the next valid random number.
+            auto result = this->state.result[this->state.position++];
```
must be 

```
auto result = state.result[0];
            state.position++
```

else we required local memory on nvidia devices

psychocoderHPC commented on include/alpaka/rand/concepts.hpp:
```diff
@@ -0,0 +1,27 @@
+ * SPDX-License-Identifier: MPL-2.0
+ */
+#pragma once
+#include "alpaka/concepts.hpp"
```
#include "alpaka/Vec.hpp"

psychocoderHPC commented on tests/unit/rand/uniformReal.cpp:
```diff
@@ -0,0 +1,227 @@
+};
+
+template<typename T_Interval>
+void IntervalChecks(auto const& x, auto const& minF, auto const& maxF)
```
`intervalChecks`

psychocoderHPC commented on tests/unit/rand/uniformReal.cpp:
```diff
@@ -0,0 +1,227 @@
+        IntervalChecks<T_Interval>(x, minF, maxF);
+        //
+        // // also check finite
+        // CHECK(std::isfinite(x));
```
dead code

psychocoderHPC commented on tests/unit/rand/uniformReal.cpp:
```diff
@@ -0,0 +1,227 @@
+
+// recursive case
+template<typename Head, typename... Tail, typename F>
+constexpr void for_each(std::tuple<Head, Tail...>, F&& f)
```
`forEach`

psychocoderHPC commented on tests/unit/rand/uniformReal.cpp:
```diff
@@ -0,0 +1,227 @@
+    constexpr auto frameExtent = 256u;
+    constexpr auto numFrames = static_cast<decltype(frameExtent)>(N / frameExtent);
+    auto hostInput = onHost::alloc<FP>(device, Vec{N});
+    for(int i = 0; i < N; i++)
```
you can use the iterator from the buffer `for(auto foo: buffer)`

TimHanel00 committed:
Addressing requested changes to the PR.
Made tests more robust (concrete value checks for edge cases) + removed a bug in the test case where the engine state is never modified.
Add another post scaling correction condition (now applying for open upper and lower bounds criteria).

psychocoderHPC committed:
small refactorings

psychocoderHPC commented:
I reviewed the pr and will push tomorrow small changes. Additionally I need to check why all non host api's run with the test into a segfault.
Mist likely a hostbuffer which is accessed on device.

psychocoderHPC committed:
fix that host memory is created on the device
